### PR TITLE
UCP: improve UT coverage of package `planner`, part 1

### LINKED ISSUE #15681: UCP: improve UT coverage of package `planner`, part 1
## Description

As the title said, this task is aimed to improve the unit test coverage of package `planner`. You can generate the coverage report and view it in browser with the following commands:

```sh
cd planner
go test -coverprofile=coverage.out # runs the unit test, writes the coverage profile to coverage.out
go tool cover -html=coverage.out   # displays the coverage report in your browser
```

The following codes should be covered:
- [ ] [planner/optimize.go#L78](https://codecov.io/gh/pingcap/tidb/src/36e89f8220e8618ea81430122a06ebc4bf04de3b/planner/optimize.go#L78)
- [ ] [planner/optimize.go#L89](https://codecov.io/gh/pingcap/tidb/src/36e89f8220e8618ea81430122a06ebc4bf04de3b/planner/optimize.go#L89)
- [ ] [planner/optimize.go#L340](https://codecov.io/gh/pingcap/tidb/src/36e89f8220e8618ea81430122a06ebc4bf04de3b/planner/optimize.go#L340)
- [ ] [planner/optimize.go#L389](https://codecov.io/gh/pingcap/tidb/src/36e89f8220e8618ea81430122a06ebc4bf04de3b/planner/optimize.go#L389)

## Score

* 100

## Mentor(s)

* @winoros 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* Basic TiDB SQL Optimization

## Learning Materials

* Chinese: [TiDB 精选技术讲解文章](https://github.com/pingcap/presentations/blob/master/hackathon-2019/reference-document-of-hackathon-2019.md)
* English: [Awesome-Database-Learning](https://github.com/pingcap/awesome-database-learning)

### PR DESCRIPTION
**Pull request by hsqlu**
Signed-off-by: Qiannan Lyu <lvqiannan@gmail.com>

UCP: #15681 

### What problem does this PR solve?

Issue Number: close #15681 
Problem Summary:

### What is changed and how it works?

What's Changed: Added more test cases

How it Works:

Improved code test coverage.

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Comment by hsqlu** [comment]
@winoros  Can you have a look please?

**Comment by zz-jason** [comment]
Could you post the coverage improvement here?

File: session/session_test.go [hunk 1]
```diff
@@ -3070,6 +3079,8 @@ func (s *testSessionSuite2) TestStmtHints(c *C) {
 	val = int64(0)
 	c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings(), HasLen, 1)
 	c.Assert(tk.Se.GetSessionVars().StmtCtx.MemTracker.CheckBytesLimit(val), IsTrue)
+	tk.MustExec("select /*+ MEMORY_QUOTA(-1 GB) */ 1;")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings(), HasLen, 1)
```
Comment (zz-jason) [inline session/session_test.go:3083]: please also check the warning message.

File: session/session_test.go [hunk 2]
```diff
@@ -2945,12 +2945,21 @@ func (s *testSessionSuite2) TestTxnGoString(c *C) {
...
+	tk.Se, err = session.CreateSession4Test(s.store)
+	c.Assert(err, IsNil)
 
+	tk.MustExec("use test")
 	tk.MustExec("create table MaxExecTime( id int,name varchar(128),age int);")
 	tk.MustExec("begin")
 	tk.MustExec("insert into MaxExecTime (id,name,age) values (1,'john',18),(2,'lary',19),(3,'lily',18);")
 
+	tk.MustQuery("select /*+ MAX_EXECUTION_TIME(1000) MAX_EXECUTION_TIME(500) */ * FROM MaxExecTime;")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings(), HasLen, 1)
```
Comment (zz-jason) [inline session/session_test.go:2959]: ditto

File: session/session_test.go (hunk 1 above)
Comment (hsqlu) [reply session/session_test.go]: The `MEMORY_QUOTA` value cannot be negative, so L340 cannot be accessed?

```
"MEMORY_QUOTA" '(' QueryBlockOpt hintIntLit UnitOfBytes ')'
	{
		maxValue := uint64(math.MaxInt64) / $5
		if $4 <= maxValue {
			$$ = &ast.TableOptimizerHint{
				HintName: model.NewCIStr($1),
				HintData: int64($4 * $5),
				QBName:   model.NewCIStr($3),
			}
		} else {
			yylex.AppendError(ErrWarnMemoryQuotaOverflow.GenWithStackByArgs(math.MaxInt64))
			parser.lastErrorAsWarn()
			$$ = nil
		}
	}
```

**Comment by hsqlu** [comment]
> Could you post the coverage improvement here?

Improvement can't be shown from the coverage percentage :(

File: bindinfo/bind_test.go [hunk 3]
```diff
@@ -432,6 +432,26 @@ func (s *testSuite) TestBindingSymbolList(c *C) {
...
+	// before binding
+	tk.MustQuery("select a, b from t where a = 3 limit 1, 100")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.IndexNames[0], Equals, "t:ia")
+	c.Assert(tk.MustUseIndex("select a, b from t where a = 3 limit 1, 100", "ia(a)"), IsTrue)
+
+	tk.MustExec(`create global binding for select a, b from t where a = 1 limit 0, 1 using select a, b from t use index (ia) where a = 1 limit 0, 1`)
+
+	tk.MustQuery("select a, b from t where a = 3 limit 1, 100")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.IndexNames[0], Equals, "t:ia")
+	c.Assert(tk.MustUseIndex("select a, b from t where a = 3 limit 1, 100", "ia(a)"), IsTrue)
```
Comment (SunRunAway) [inline bindinfo/bind_test.go:452]: Better to add one more test about index `b`.

**Comment by hsqlu** [comment]
@winoros @SunRunAway PTAL

**Comment by zz-jason** [comment]
> > Could you post the coverage improvement here?
> 
> Improvement can't be shown from the coverage percentage :(

You can use the following method to generate the UT coverage and upload the coverage file to the PR description:

```sh
cd planner
go test -coverprofile=coverage.out # runs the unit test, writes the coverage profile to coverage.out
go tool cover -html=coverage.out   # displays the coverage report in your browser
```

**Comment by hsqlu** [comment]
@zz-jason PTAL.

The command you mentioned doesn't work.
```
➜  planner git:(master) go test -coverprofile=coverage.out
?       github.com/pingcap/tidb/planner [no test files]
```

I changed it to below. Optimize() was improved from 92.1% to 96.8% by `bind_test.go`

```
➜  bindinfo git:(master) go test -coverpkg=../planner -coverprofile=coverage.out ./bind_test.go
ok      command-line-arguments  1.289s  coverage: 73.3% of statements in ../planner
➜  bindinfo git:(master) go tool cover -func=coverage.out                                      
github.com/pingcap/tidb/planner/optimize.go:41:         Optimize                        92.1%
github.com/pingcap/tidb/planner/optimize.go:143:        optimize                        85.2%
github.com/pingcap/tidb/planner/optimize.go:193:        extractSelectAndNormalizeDigest 100.0%
github.com/pingcap/tidb/planner/optimize.go:213:        getBindRecord                   95.0%
github.com/pingcap/tidb/planner/optimize.go:244:        handleInvalidBindRecord         75.0%
github.com/pingcap/tidb/planner/optimize.go:258:        handleEvolveTasks               85.7%
github.com/pingcap/tidb/planner/optimize.go:277:        useMaxTS                        100.0%
github.com/pingcap/tidb/planner/optimize.go:289:        OptimizeExecStmt                0.0%
github.com/pingcap/tidb/planner/optimize.go:305:        handleStmtHints                 43.1%
github.com/pingcap/tidb/planner/optimize.go:397:        init                            100.0%
total:                                                  (statements)                    73.3%
➜  bindinfo git:(master) gco UCP-improve-UT-planner-1    
Switched to branch 'UCP-improve-UT-planner-1'
➜  bindinfo git:(UCP-improve-UT-planner-1) go test -coverpkg=../planner -coverprofile=coverage.out ./bind_test.go
ok      command-line-arguments  1.246s  coverage: 74.8% of statements in ../planner
➜  bindinfo git:(UCP-improve-UT-planner-1) go tool cover -func=coverage.out                                      
github.com/pingcap/tidb/planner/optimize.go:41:         Optimize                        96.8%
github.com/pingcap/tidb/planner/optimize.go:143:        optimize                        85.2%
github.com/pingcap/tidb/planner/optimize.go:193:        extractSelectAndNormalizeDigest 100.0%
github.com/pingcap/tidb/planner/optimize.go:213:        getBindRecord                   95.0%
github.com/pingcap/tidb/planner/optimize.go:244:        handleInvalidBindRecord         75.0%
github.com/pingcap/tidb/planner/optimize.go:258:        handleEvolveTasks               85.7%
github.com/pingcap/tidb/planner/optimize.go:277:        useMaxTS                        100.0%
github.com/pingcap/tidb/planner/optimize.go:289:        OptimizeExecStmt                0.0%
github.com/pingcap/tidb/planner/optimize.go:305:        handleStmtHints                 43.1%
github.com/pingcap/tidb/planner/optimize.go:397:        init                            100.0%
total:                                                  (statements)                    74.8%
```


And coverage for handleStmtHints was improved from 93.1% to 96.6%.

```
➜  session git:(master) go tool cover -func=coverage.out                             
github.com/pingcap/tidb/planner/optimize.go:41:         Optimize                        36.5%
github.com/pingcap/tidb/planner/optimize.go:143:        optimize                        96.3%
github.com/pingcap/tidb/planner/optimize.go:193:        extractSelectAndNormalizeDigest 100.0%
github.com/pingcap/tidb/planner/optimize.go:213:        getBindRecord                   80.0%
github.com/pingcap/tidb/planner/optimize.go:244:        handleInvalidBindRecord         0.0%
github.com/pingcap/tidb/planner/optimize.go:258:        handleEvolveTasks               0.0%
github.com/pingcap/tidb/planner/optimize.go:277:        useMaxTS                        100.0%
github.com/pingcap/tidb/planner/optimize.go:289:        OptimizeExecStmt                0.0%
github.com/pingcap/tidb/planner/optimize.go:305:        handleStmtHints                 93.1%
github.com/pingcap/tidb/planner/optimize.go:397:        init                            100.0%
total:                                                  (statements)                    64.8%
```

```
➜  session git:(UCP-improve-UT-planner-1) go tool cover -func=coverage.out                             
github.com/pingcap/tidb/planner/optimize.go:41:         Optimize                        36.5%
github.com/pingcap/tidb/planner/optimize.go:143:        optimize                        96.3%
github.com/pingcap/tidb/planner/optimize.go:193:        extractSelectAndNormalizeDigest 100.0%
github.com/pingcap/tidb/planner/optimize.go:213:        getBindRecord                   80.0%
github.com/pingcap/tidb/planner/optimize.go:244:        handleInvalidBindRecord         0.0%
github.com/pingcap/tidb/planner/optimize.go:258:        handleEvolveTasks               0.0%
github.com/pingcap/tidb/planner/optimize.go:277:        useMaxTS                        100.0%
github.com/pingcap/tidb/planner/optimize.go:289:        OptimizeExecStmt                0.0%
github.com/pingcap/tidb/planner/optimize.go:305:        handleStmtHints                 96.6%
github.com/pingcap/tidb/planner/optimize.go:397:        init                            100.0%
total:                                                  (statements)                    65.7%
```

**Pull request merged**